File: field/field2dyna.m
function [dynaImat] = field2dyna(NodeName, field_params_json, ElemName, ForceNonlinear)
% function [dynaImat] = field2dyna(NodeName, field_params_json, ElemName, ForceNonlinear)
%
% INPUT:
%   NodeName (string) - file name to read nodes from (e.g., nodes.dyn); needs to
%                       be a comma-delimited file with header/footer lines that
%                       start with '*'
%   field_params_json (str) - JSON filename
%   ElemName (string) - file name to read elements from (default: elems.dyn);
%                       like node file, needs to be comma-delimited.
%   ForceNonlinear(int) - optional input argument. Set as 1 if you want to
%                         force nodal volumes to be calculated.
%
% OUTPUT:
%   dyna-I-*.mat file is saved to CWD and filename string returned
%
% EXAMPLE:
% dynaImatfile = field2dyna('nodes.dyn', 0.5, 1.3, [0 0 0.02], 7.2, 'vf105', 'gaussian', ...
%                            12, 'elems.dyn', 1);
%
% Mark Palmeri
% mlp6@duke.edu
% 2015-03-17

measurementPointsandNodes = read_mpn(NodeName);

% check to see if there are nodes in the x = y = 0 plane to make sure that
% "on axis" intensities are properly captured
check_on_axis(measurementPointsandNodes(:,2:4));

% invert the z axis
measurementPointsandNodes(:,4)=-measurementPointsandNodes(:,4);

% switch x and y so plane of symmetry is elevation, not lateral
measurementPointsandNodes(:,2:3)=[measurementPointsandNodes(:,3) measurementPointsandNodes(:,2)];
  
% convert from cm -> m
measurementPointsandNodes(:,2:4)=measurementPointsandNodes(:,2:4)/100;

FIELD_PARAMS = read_json(field_params_json);

FIELD_PARAMS.measurementPointsandNodes = measurementPointsandNodes;
% setup some input argument defaults
if (nargin < 8),
    threads = 1;
end;

if (nargin < 9),
    lownslow = true;
end;

% perform the field calculation
[intensity, FIELD_PARAMS] = dynaField(FIELD_PARAMS);

% save intensity file
dynaImat = sprintf('dyna-I-f%.2f-F%.1f-FD%.3f-a%.2f.mat', ...
                   FIELD_PARAMS.freq_MHz, FIELD_PARAMS.fnum, ...
                   FIELD_PARAMS.focus_m(3), FIELD_PARAMS.alpha_dB_cm_MHz);
save(dynaImat, 'intensity', 'FIELD_PARAMS');

% check if non-uniform force scaling must be done
isUniform = checkUniform(measurementPointsandNodes(:,2:4));
if (nargin < 4)
    ForceNonlinear = 0;
end
if (~isUniform || ForceNonlinear == 1)
    % should run calcNodeVol.py
    fprintf('This is a non-linear mesh. Generating node volume file.\n')
    if (nargin < 11)
        warning('This is a non-linear mesh, but elements file was not given as an input argument. Skipping node volume file generation.')
    else
        NodeVolumeFilename = ['NodeVolume_' NodeName '_' ElemName '.txt'];
        if (exist(NodeVolumeFilename, 'file') == 0)
            eval(sprintf('nodeVolSuccess = system(''python calcNodeVol.py --nodefile %s --elefile %s --nodevolfile %s'');', NodeName, ElemName, NodeVolumeFilename));
            if (nodeVolSuccess ~= 0)
              fprintf('Node volume generation failed. Check to make sure node and element files exist in specified directory.\n')
            else
              fprintf('%s has been created.\n', ['NodeVolume_' NodeName '_' ElemName '.txt'])
            end
        else
            disp('Node volume file already exists; not generating new one');
        end
    end
else
    fprintf('This is a linear mesh.\n');
end

disp('The next step is to run makeLoadsTemps.');
disp('This will generate point loads and initial temperatures.');

File: fieldC/field2dyna.c
/*
% function [dynaImat] = field2dyna(NodeName, alpha, Fnum, focus, Frequency,
%	Transducer, Impulse, threads, lownslow, ElemName, verbose)
%
% INPUT:
%   NodeName (string) - file name to read nodes from (e.g., nodes.dyn); needs
%                       to be a comma-delimited file with header/footer lines
%                       that start with '*'
%   alpha_dBcmMHz - 0.5, 1.0, etc. (dB/cm/MHz)
%   Fnum - F/# (e.g. 1.3)
%   focusM - [x y z] (meters) "Field" coordinates
%   Frequency - excitation frequency (MHz)
%   Transducer (string) - 'vf105','vf73'
%   Impulse (string) - 'gaussian','exp'
%   threads (int) - number of parallel threads to use for Field II Pro
%   lownslow (bool) - low memory footprint, or faster parallel (high RAM)
%		[default = 0]
%   ElemName (string) - file name to read elements from (default: elems.dyn);
%                       like node file, needs to be comma-delimited.
%
% OUTPUT:
%   dyna-I-*.mat file is saved to CWD and filename string returned
%
% EXAMPLE:
% dynaImatfile = field2dyna('nodes.dyn', 0.5, 1.3, [0 0 0.02], 7.2, 'vf105',
%	'gaussian', ..., 12, 'elems.dyn', 1);
%
% Mark Palmeri
% mlp6@duke.edu
% 2015-03-17
%
* Ned Danieley
*
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "field.h"

int checkOnAxis();

char *
field2dyna(char *nodeName, double alpha_dBcmMHz, double fnum, point_type focusM,
	double freqMHz, char *transducer, char *impulse, int threads, int lowNslow,
	char *elemName, int verbose)
{
int dynaField();
int i, numNodes;
int isUniform;
double temp;
struct nodeEntry *pointsAndNodes, *readMpn();
struct FieldParams fieldParams;
char nodeVolFileName[80];
char calcNodeVolCmd[80];
int status;

	fprintf(stderr, "in field2dyna, verbose %d\n", verbose);

	if (verbose >= 1) {
		fprintf(stderr, "in field2dyna, calling readMpn; node name %s\n",
			nodeName);
		fprintf(stderr, "in field2dyna, alpha_dBcmMHz %f fnum %f freqMHz %f\n",
			alpha_dBcmMHz, fnum, freqMHz);
		fprintf(stderr, "in field2dyna, focus x %f y %f z %f \n", focusM.x,
			focusM.y, focusM.z);
		fprintf(stderr, "in field2dyna, transducer %s\n", transducer);
		fprintf(stderr, "in field2dyna, impulse %s\n", impulse);
		fprintf(stderr, "in field2dyna, threads %d\n", threads);
		fprintf(stderr, "in field2dyna, lowNslow %d\n", lowNslow);
		fprintf(stderr, "in field2dyna, element name %s\n", elemName);
		fprintf(stderr, "in field2dyna, verbose %d\n", verbose);
		}

	pointsAndNodes = readMpn(nodeName, &numNodes, verbose);
	if (pointsAndNodes == NULL) {
		fprintf(stderr, "didn't get enough values from readMpn\n");
		exit(0);
		}

	if (verbose >= 1) fprintf(stderr, "after readMpn; numNodes %d\n", numNodes);

	if (verbose == 3) for (i = 0; i < 13; i++)
	    fprintf(stderr, "field2dyna 1, node %d is %d, %f, %f, %f\n", i, pointsAndNodes[i].nodeID, pointsAndNodes[i].x, pointsAndNodes[i].y, pointsAndNodes[i].z);

	if (!checkOnAxis(pointsAndNodes, numNodes)) {
		fprintf(stderr, "There are no nodes in the lateral / elevation plane = 0 (imaging plane).\n");
		fprintf(stderr, "This can lead to inaccurate representations of the intensity fields!!')\n");
		}

	if (verbose == 3) for (i = 0; i < 13; i++)
	    fprintf(stderr, "field2dyna 2, node %d is %d, %f, %f, %f\n", i, pointsAndNodes[i].nodeID, pointsAndNodes[i].x, pointsAndNodes[i].y, pointsAndNodes[i].z);

/* invert the z axis */
	for (i = 0; i < numNodes; i++) pointsAndNodes[i].z = -pointsAndNodes[i].z;

	if (verbose == 3) for (i = 0; i < 13; i++)
	    fprintf(stderr, "field2dyna 3, node %d is %d, %f, %f, %f\n", i, pointsAndNodes[i].nodeID, pointsAndNodes[i].x, pointsAndNodes[i].y, pointsAndNodes[i].z);

/* switch x and y */
	for (i = 0; i < numNodes; i++) {
		temp = pointsAndNodes[i].x;
		pointsAndNodes[i].x = pointsAndNodes[i].y;
		pointsAndNodes[i].y = temp;
		}
	if (verbose == 3) for (i = 0; i < 13; i++)
	    fprintf(stderr, "field2dyna 4, node %d is %d, %f, %f, %f\n", i, pointsAndNodes[i].nodeID, pointsAndNodes[i].x, pointsAndNodes[i].y, pointsAndNodes[i].z);


/* change from centimeters to meters */
	for (i = 0; i < numNodes; i++) {
		pointsAndNodes[i].x /= 100;
		pointsAndNodes[i].y /= 100;
		pointsAndNodes[i].z /= 100;
		}

	if (verbose == 3) for (i = 0; i < 13; i++)
	    fprintf(stderr, "field2dyna 5, node %d is %d, %f, %f, %f\n", i, pointsAndNodes[i].nodeID, pointsAndNodes[i].x, pointsAndNodes[i].y, pointsAndNodes[i].z);

/* populate structure for dynaField */

	fieldParams.pointsAndNodes = pointsAndNodes;
	fieldParams.alpha_dBcmMHz = alpha_dBcmMHz;
	fieldParams.fnum = fnum;
	fieldParams.focusM = focusM;
	fieldParams.frequencyMHz = freqMHz;
	fieldParams.transducer = transducer;
	fieldParams.impulse = impulse;
	fieldParams.soundSpeed_MperSec = 1540;
	fieldParams.samplingFrequencyHz = 100e6;
	fieldParams.threads = threads;

	if (verbose >= 1) fprintf(stderr, "in field2dyna, focus x %f y %f z %f\n", fieldParams.focusM.x, fieldParams.focusM.y, fieldParams.focusM.z);

/* call dynaField here */

	if (dynaField(fieldParams, threads, numNodes, lowNslow, verbose) == 0) {
		fprintf(stderr, "in field2dyna, call to dynaField failed\n");
		exit(0);
		}
}
